File: scala/src/main/scala/day02/Day02.scala
package day02

import scala.annotation.tailrec
import scala.io.Source

enum Play:
  case ROCK, PAPER, SCISSORS

enum Result:
  case LOSE, DRAW, WIN

def toPlay(char: String): Play  =
  char match {
    case "X" => Play.ROCK
    case "A" => Play.ROCK
    case "Y" => Play.PAPER
    case "B" => Play.PAPER
    case "Z" => Play.SCISSORS
    case "C" => Play.SCISSORS
    case _ =>  throw new RuntimeException("Should never happen")
  }
def toResult(char: String): Result  =
  char match
    case "X" => Result.LOSE
    case "Y" => Result.DRAW
    case "Z" => Result.WIN
    case _ =>  throw new RuntimeException("Should never happen")

def score(game: (Play, Play)) : Int =
  game match
    case (Play.PAPER, Play.ROCK) => 1
    case (Play.SCISSORS, Play.PAPER) => 2
    case (Play.ROCK, Play.SCISSORS) => 3

    case (Play.ROCK, Play.ROCK) => 1 + 3
    case (Play.PAPER, Play.PAPER) => 2 + 3
    case (Play.SCISSORS, Play.SCISSORS) => 3 + 3

    case (Play.SCISSORS, Play.ROCK) => 1 + 6
    case (Play.ROCK, Play.PAPER) => 2 + 6
    case (Play.PAPER, Play.SCISSORS) => 3 + 6

def score2(game: (Play, Result)): Int =
  game match
    case (Play.PAPER, Result.LOSE) => 1
    case (Play.PAPER, Result.DRAW) => 5
    case (Play.PAPER, Result.WIN) => 9
    case (Play.ROCK, Result.LOSE) => 3
    case (Play.ROCK, Result.DRAW) => 4
    case (Play.ROCK, Result.WIN) => 8
    case (Play.SCISSORS, Result.LOSE) => 2
    case (Play.SCISSORS, Result.DRAW) => 6
    case (Play.SCISSORS, Result.WIN) => 7

def parseInput(line: String): (Play, Play) =
  line match {
    case s"$opponent $me" => (toPlay(opponent), toPlay(me))
    case _ => throw new RuntimeException(s"Not matching line: $line")
  }
def parseInput2(line: String): (Play, Result) =
  line match {
    case s"$opponent $me" => (toPlay(opponent), toResult(me))
    case _ => throw new RuntimeException(s"Not matching line: $line")
  }


@main
def main(): Unit =
  val input = Source.fromResource("day02.txt").getLines().toList
  val rounds = input
    .map(parseInput)
    .map(score)
    .sum

  println("Pt1: " + rounds)

  val rounds2 = input
    .map(parseInput2)
    .map(score2)
    .sum
  println("Pt2: " + rounds2)

File: rust/src/day02.rs
enum Play { ROCK, PAPER, SCISSORS }

enum Result { WIN, DRAW, LOSE }

impl Result {
    fn score(&self) -> u32 {
        match *self {
            Result::WIN => 0,
            Result::DRAW => 3,
            Result::LOSE => 6,
        }
    }
}

impl Play {
    fn score(&self) -> u32 {
        match *self {
            Play::ROCK => 1,
            Play::PAPER => 2,
            Play::SCISSORS => 3,
        }
    }
}

fn result(opponent: &Play, me: &Play) -> Result {
    match opponent {
        Play::ROCK =>
            match me {
                Play::ROCK => Result::DRAW,
                Play::PAPER => Result::LOSE,
                Play::SCISSORS => Result::WIN,
            }
        Play::PAPER =>
            match me {
                Play::ROCK => Result::WIN,
                Play::PAPER => Result::DRAW,
                Play::SCISSORS => Result::LOSE,
            }
        Play::SCISSORS =>
            match me {
                Play::ROCK => Result::LOSE,
                Play::PAPER => Result::WIN,
                Play::SCISSORS => Result::DRAW,
            }
    }
}

fn pt1(strategy_guide: &str) -> u32 {
    strategy_guide
        .lines()
        .map(|l| parse_play(l))
        .map(|g| score_game(g))
        .sum()
}

fn score_game((opponent, me): (Play, Play)) -> u32 {
    me.score() + result(&opponent, &me).score()
}

fn parse_play(line: &str) -> (Play, Play) {
    line
        .trim()
        .split_once(' ')
        .map(|(s1, s2)| (string_to_play(&s1), string_to_play(&s2)))
        .unwrap()
}

fn string_to_play(s: &str) -> Play {
    match s {
        "A" => Play::ROCK,
        "B" => Play::PAPER,
        "C" => Play::SCISSORS,
        "X" => Play::ROCK,
        "Y" => Play::PAPER,
        "Z" => Play::SCISSORS,
        _ => panic!("No matching play!")
    }
}

pub fn day02() {
    let input = include_str!("../input/day02.txt");
    println!("Day 2, part 1: {}", pt1(input));
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_input_pt1() {
        let input = "A Y
                           B X
                           C Z";

        assert_eq!(pt1(input), 15)
    }
}

